endpoints to receive/forward hq invites

sravfeyn opened the pull request:
[Spec](https://docs.google.com/document/d/1h9OlQDD7Jpygwyi-Zp6aU-_TQNkz4wOxw93ZOjkTwHY/edit?tab=t.0#heading=h.syj1ugk3stxe)
[Ticket](https://dimagi.atlassian.net/browse/CCCT-425)

calellowitz commented on users/views.py:
```diff
@@ -417,6 +419,52 @@ def post(self, request, *args, **kwargs):
+    authentication_classes = [ClientProtectedResourceAuth]
+
+    def post(self, request, *args, **kwargs):
+        callback_url = quote_plus(request.data["callback_url"])
```
where does `callback_url` come from and point to?

sravfeyn replied:
This is all according to the modified spec.

This comes from HQ and it points to HQ's invite confirmation view. It's for actual invite confirmation, when mobile confirms it to ConnectID, ConnectID calls back HQ on callback_url with the user's token.

sravfeyn replied:
This is the HQ change https://github.com/dimagi/commcare-hq/pull/34967/commits/407303f6c68034fd182f7325a2764eb3f608e74b

calellowitz commented on users/views.py:
```diff
@@ -417,6 +419,52 @@ def post(self, request, *args, **kwargs):
+
+
+class ConfirmHQInviteCallback(APIView):
+    authentication_classes = [ClientProtectedResourceAuth]
```
Who makes this call? Is this not from mobile?

sravfeyn replied:
Yes, this is made by mobile.

# From spec
To accept invitation (when user confirms), call below endpoint

‘connectid.dimagi.com/users/confirm_hq_invite’

POST params
callback_url: callback_url (forwarded in deep link)
invite_code: invite_code (forwarded in deep link)
user_token: user’s active auth token

calellowitz replied:
In that case, this auth is incorrect. It only works with oauth client credentials grants (what we use for connect or other servers). I think the default auth classes should be fine.

sravfeyn replied:
You are right! I will remove those classes.

calellowitz reviewed (commented):
Couple small questions

sravfeyn commented:
@calellowitz Please let me know if this is good to go.

calellowitz commented on users/views.py:
```diff
@@ -417,6 +419,52 @@
+        user_token = request.data["user_token"]
+        invite_code = unquote_plus(callback_url)
+        try:
+            response = requests.post(callback_url, data={"invite_code": invite_code, "token": user_token})
```
Are you planning to address this? Given your answer below about where this points, it should be doable to check it against an allowed list (only hq instances?)

calellowitz commented on connectid/views.py:
```diff
@@ -28,3 +28,10 @@ def assetlinks_json(request):
     return JsonResponse(assetfile, safe=False)
+
+
+def deeplink(request, subpath):
```
Are we able to direct them directly to the play store? That is what we discussed on the call and seems to make the most sense if it is possible. Mobile seemed to believe that it would be.

sravfeyn replied:
Right, I have updated it to redirect to the app.

<img width="1146" alt="Screenshot 2024-11-22 at 12 31 38 PM" src="https://github.com/user-attachments/assets/223ee5de-018d-45d3-9d9b-e98e9bd0229c">

calellowitz commented on users/views.py:
```diff
@@ -434,7 +434,8 @@ def post(self, request, *args, **kwargs):
             # So fake a success message
             return JsonResponse({"success": True})
-        deeplink = f"https://connectid.dimagi.com/hq_invite/{callback_url}/{username}/{invite_code}/{user_domain}"
+        details = f"{callback_url}/{username}/{invite_code}/{user_domain}/{user.username}"
```
I think we spoke about this a while ago, but why is this encoded as a path rather than using query params (which libraries on web and mobile could parse much more easily)?

sravfeyn replied:
Yeah, I think there was some issue regarding formatting this using query params on mobile, this format was specified by mobile.

calellowitz replied:
I spoke to @OrangeAndGreen and he seemed to believe query params would be better for them as well. Let's offline on slack about this with a mobile dev.

sravfeyn replied:
Okay, I have updated this to use queryparams.

calellowitz reviewed (changes requested):
Left a few questions on the new code in addition to the previous comments

sravfeyn commented:
@calellowitz Thanks for the review, I have followed up.

calellowitz commented on connectid/views.py:
```diff
@@ -31,7 +31,42 @@ def assetlinks_json(request):
-    #   and the URL is opened in the browser, ask the
-    #   user to install the app and try again
-    return HttpResponse("Please install CommCare and re-open this link to accept invite")
+    html_content = """
```
Why inline HTML instead of using a separate file

sravfeyn replied:
I have moved it to its own dir.

calellowitz commented on connectid/settings.py:
```diff
@@ -80,7 +80,10 @@
 
-
+TRUSTED_COMMCAREHQ_HOSTS = [
+    "commcarehq.org",
```
how does this work? Specifically, if `commcarehq.org` is enough to catch `www.commcarehq.org` I would expect it to also catch staging, and if we need to be more specific, I would expect `www` to be required.

sravfeyn replied:
Good point, I will add `www` to the list to be specific.

sravfeyn commented:
@calellowitz I have addressed your feedback, please have a look. I think the CodeQL failure is incorrect as it is not able to detect the trusted URL check that is added.

sravfeyn commented:
@calellowitz Bumping for review please.

sravfeyn commented:
Thanks @calellowitz please also review the corresponding HQ PR https://github.com/dimagi/commcare-hq/pull/34967/files#diff-dc2c79b556e950acc7dea90175fd5febba22bf9fa4faaba1b61c3b3d5acdc9e9
